File: R package/src/cov.c
/* ID: cov_weighted.c, last updated 2020-09-23, F.Osorio */

#include "fastmatrix.h"

void
cov_weighted(double *x, int *nobs, int *vars, double *weights, double *mean, double *cov)
{ /* wrapper for 'FM_mean_and_Scatter' */
  int n = *nobs, p = *vars;

  FM_center_and_Scatter(x, n, p, weights, mean, cov);
}

void
cov_MSSD(double *x, int *nobs, int *vars, double *mean, double *cov)
{ /* wrapper for 'FM_cov_MSSD' */
  int n = *nobs, p = *vars;

  FM_cov_MSSD(x, n, p, mean, cov);
}

File: pkg/R/cov.R
## ID: cov.weighted.R, last updated 2020-09-23, F.Osorio

cov.weighted <- function(x, weights = rep(1, nrow(x)))
{ ## computes weighted mean and covariance using an online algorithm
  if (is.data.frame(x))
    x <- as.matrix(x)
  else if (!is.matrix(x))
    stop("'x' must be a matrix or a data frame")
  if (!all(is.finite(x)))
    stop("'x' must contain finite values only")
  n <- nrow(x)
  p <- ncol(x)
  if (!missing(weights)) {
    if (length(weights) != n)
      stop("length of 'weights' must equal the number of rows in 'x'")
    if (any(weights < 0) || (sum(weights) == 0))
	    stop("weights must be non-negative and not all zero")
  }

  storage.mode(x) <- "double"
  y <- .C("cov_weighted",
          x = x,
          n = as.integer(n),
          p = as.integer(p),
          weights = as.double(weights),
          mean = double(p),
          cov  = double(p * p))[c("mean","cov")]
  y$cov <- matrix(y$cov, nrow = p, ncol = p)
  y
}

cov.MSSD <- function(x)
{ ## covariance estimation using the Mean Square Successive Method (MSSD)
  if (is.data.frame(x))
    x <- as.matrix(x)
  else if (!is.matrix(x))
    stop("'x' must be a matrix or a data frame")
  if (!all(is.finite(x)))
    stop("'x' must contain finite values only")
  n <- nrow(x)
  p <- ncol(x)

  storage.mode(x) <- "double"
  y <- .C("cov_MSSD",
          x = x,
          n = as.integer(n),
          p = as.integer(p),
          mean = double(p),
          cov  = double(p * p))[c("mean","cov")]
  y$cov <- matrix(y$cov, nrow = p, ncol = p)
  y
}
